### PR TITLE
Show error pages if user isn't in proper session state to view them

### PR DESCRIPTION
**Pull request by toolness**
This decorates a bunch of our letter builder steps with various checks that ensure the user has a session state in which it actually makes sense for the user to see them.  If it doesn't make sense, we show alternate content with a call to action that points them in the right direction.

Broadly speaking we have a few kinds of steps:

* **Onboarding steps**, where the user is entering information before they've actually created an account.  In these steps it's actually _bad_ for a user to be logged-in, so we want to direct them to the latest step they've completed.

* **Not sent letter steps**, where the user has created an account but not yet sent a rent letter for the current rent period.  For these steps, if the user isn't logged-in, we want to direct them to log in; and if they are logged in but have already sent a letter, we want to direct them to the confirmation page.

Aside from that there are a few other edge cases which this PR attempts to handle.

### CODE REVIEW

**Comment by toolness** [comment]
This PR reduces the overall test coverage because our smoke tests no longer cause the code of all our pages to be trivially executed (their execution is blocked by e.g. the user not being logged in).  Rather than addressing them now, I'd like to fix things in #1226.

**Comment by toolness** [comment]
[ch3432]
[ch3433]
[ch3434]

File: frontend/lib/norent/letter-builder/error-pages.tsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  children: React.ReactNode;
+};
+
+const CtaLink: React.FC<CtaLinkProps> = (props) => (
+  <p className="has-text-centered">
+    <Link className="button is-primary is-large jf-is-extra-wide" to={props.to}>
+      {props.children}
+    </Link>
+  </p>
+);
```
Comment (sraby) [inline frontend/lib/norent/letter-builder/error-pages.tsx:23]: Hey @toolness, have you considered using the `<CenteredPrimaryButtonLink>` button component here? Looks almost identical to what you've implemented here.

File: frontend/lib/util/session-predicates.ts [hunk 2]
```diff
@@ -4,3 +4,8 @@ import { AllSessionInfo } from "../queries/AllSessionInfo";
 export function isUserLoggedIn(s: AllSessionInfo): boolean {
   return !!s.phoneNumber;
 }
```
Comment (sraby) [inline frontend/lib/util/session-predicates.ts:6]: Ooooh! This is the first time I've truly noticed this "double bang" js syntax— super helpful to keep in mind for returning truthy as a boolean :)

File: frontend/lib/ui/session-error-handling.tsx [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+};
+
+/**
+ * Wraps the given component in error-handling logic, such
+ * that if the current user's session matches a certain
+ * error state, a error component is shown instead of the
+ * usual one.
+ */
+export function withSessionErrorHandling<T>(
+  isInErrorState: NullishSessionPredicate,
```
Comment (sraby) [inline frontend/lib/ui/session-error-handling.tsx:42]: Haha, I think since you have a prop called `NullishSessionPredicate` I can understand how you'd call this "over-abstracted." Good learning experience for me to go through this though

File: frontend/lib/norent/letter-builder/error-pages.tsx (hunk 1 above)
Comment (toolness) [reply frontend/lib/norent/letter-builder/error-pages.tsx:23]: ohhh good point, the funny thing is that I actually pulled this code from some other existing code on a different NoRent page, so I filed #1326 to address it during cleanup!

File: frontend/lib/util/session-predicates.ts (hunk 2 above)
Comment (toolness) [reply frontend/lib/util/session-predicates.ts:6]: Ohhh yeah it is super useful although a bit annoying to read.  I keep have to mentally walk through the logic every time I see it.

File: frontend/lib/ui/session-error-handling.tsx (hunk 3 above)
Comment (toolness) [reply frontend/lib/ui/session-error-handling.tsx:42]: Hahaha yeah, I got tired of having to always return a boolean in some of the session predicates, and I figured it's not such a big deal to force a boolean here since we should only really raise an error the value is definitely true... ahhh maybe I should make the regular `SessionPredicate` something that can be nullish, i dunno.
